transformers/modeling_utils.py

```diff
@@ -87,6 +87,19 @@ def __init__(self, config, *inputs, **kwargs):
     def base_model(self):
         return getattr(self, self.base_model_prefix, self)
 
+    def decode(self, input_ids, **model_kwargs):
+        """ Decode the input sequence; preprocess the input before
+        the forward pass if necessary.
+        """
+        raise NotImplementedError("Only models with LM heads can be used as decoders")
```

Comment:
I think it's better to have a test that the model has an output_embedding (with `get_output_embeddings`) (which only models with LM heads have) and if so initialize a sampler.

The Sampler can also be initialized when loading the model but we have to be careful that the `__init__` method of the based class will be called before the creation of the embeddings by the derived classes.

So the initialization of the sampler should be done in the `init_weights()` method of `PretrainedModel`. 

The whole idea is to hide the Sampler classes from the user so the `decode` should handle the call to the sampler and the Sampler should be loaded by default when a model with an LM head is initialized.

I will help you on this.

Reply:
See, when you talked about hiding the logic I understood putting everything behind a factory function. It's unfortunate, but now I think I see. Before I start writing code again, let me try to explicit what I think you want. Please correct me if I'm wrong.

The users should be able to call the sampler as follows:

```python
prompt_ids = torch.tensor(tokenizer.encode("this is my text")).unsqeeze(0)
generated_sequence = model.decode(length=10, prompt_ids=prompt_ids, temperature=1.3, k=9, p=0.9)
```

And raises an error if the model does not have an `output_embedding`. You can stop reading here if I have the wrong idea.

My question then is: what happens in terms of API when we introduce beam search (which is coming anyway with the PR on summarization), potentially others in the future? A new method `decode_beam_search`? 

There is also the issue of carrying a stateful object in the model.  As a user I would expect that `Model.decoder(my_prompt, **sampler_kwargs)` always returns a sequence solely based on this prompt and the parameters I passed with the kwargs (k, p, temperature, etc.). Now imagine I call the method again but this time I omit the temperature. What is the expected behavior? Keep the previous temperature (as now set in the class) or going back to the default (which means re-initializing the class)? How do we make this behavior obvious for the users?

Reply:
Yes, that was the behavior I meant in my proposal.

Regarding the state, in my original proposal I mentioned adding a series of attributes to `PretrainedConfig` (see the second point in the proposal) to configure the state for the decoder, like `top_k` (int), `top_p` (float), `beams` (int with beams == 1 being equivalent to no beam search), `sampling` (boolean). We can make them overrideable by specific keyword values passed to `Model.decode()`.

Reply:
Ok, I definitely agree the UX is simpler this way. Let's see how we can make the internals work.